### PR TITLE
Fix logging JDBC SQL error when task fails

### LINKED ISSUE #16295: JDBC operator not logging errors
Hi, 

Since Airflow 2.0, we are having issues with logging for the JDBC operator. When such a tasks fails, we only see
`INFO - Task exited with return code 1`
The actual error and stack trace is not present.
It also seems to not try to execute it again, it only tries once even though my max_tries is 3.

I am using a Local Executor, and logs are also stored locally.
This issue occurs for both local installations and Docker.

full log:

```
*** Reading local file: /home/stijn/airflow/logs/airflow_incr/fmc_mtd/2021-06-01T15:00:00+00:00/1.log
[2021-06-01 18:00:13,389] {taskinstance.py:876} INFO - Dependencies all met for <TaskInstance: airflow_incr.fmc_mtd 2021-06-01T15:00:00+00:00 [queued]>
[2021-06-01 18:00:13,592] {taskinstance.py:876} INFO - Dependencies all met for <TaskInstance: airflow_incr.fmc_mtd 2021-06-01T15:00:00+00:00 [queued]>
[2021-06-01 18:00:13,592] {taskinstance.py:1067} INFO - 
--------------------------------------------------------------------------------
[2021-06-01 18:00:13,592] {taskinstance.py:1068} INFO - Starting attempt 1 of 4
[2021-06-01 18:00:13,593] {taskinstance.py:1069} INFO - 
--------------------------------------------------------------------------------
[2021-06-01 18:00:13,975] {taskinstance.py:1087} INFO - Executing <Task(JdbcOperator): fmc_mtd> on 2021-06-01T15:00:00+00:00
[2021-06-01 18:00:13,980] {standard_task_runner.py:52} INFO - Started process 957 to run task
[2021-06-01 18:00:13,983] {standard_task_runner.py:76} INFO - Running: ['airflow', 'tasks', 'run', 'airflow_incr', 'fmc_mtd', '2021-06-01T15:00:00+00:00', '--job-id', '2841', '--pool', 'default_pool', '--raw', '--subdir', 'DAGS_FOLDER/100_FL_DAG_airflow_incr_20210531_122511.py', '--cfg-path', '/tmp/tmp67h9tgso', '--error-file', '/tmp/tmp4w35rr0g']
[2021-06-01 18:00:13,990] {standard_task_runner.py:77} INFO - Job 2841: Subtask fmc_mtd
[2021-06-01 18:00:15,336] {logging_mixin.py:104} INFO - Running <TaskInstance: airflow_incr.fmc_mtd 2021-06-01T15:00:00+00:00 [running]> on host DESKTOP-VNC70B9.localdomain
[2021-06-01 18:00:17,757] {taskinstance.py:1282} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=Vaultspeed
AIRFLOW_CTX_DAG_ID=airflow_incr
AIRFLOW_CTX_TASK_ID=fmc_mtd
AIRFLOW_CTX_EXECUTION_DATE=2021-06-01T15:00:00+00:00
AIRFLOW_CTX_DAG_RUN_ID=scheduled__2021-06-01T15:00:00+00:00
[2021-06-01 18:00:17,757] {jdbc.py:70} INFO - Executing: ['INSERT INTO "moto_fmc"."fmc_loading_history" \n\t\tSELECT \n\t\t\t\'airflow_incr\',\n\t\t\t\'airflow\',\n\t\t\t35,\n\t\t\tTO_TIMESTAMP(\'2021-06-01 16:00:00.000000\', \'YYYY-MM-DD HH24:MI:SS.US\'::varchar),\n\t\t\t"fmc_begin_lw_timestamp" + -15 * interval\'1 minute\',\n\t\t\tTO_TIMESTAMP(\'2021-06-01 16:00:00.000000\', \'YYYY-MM-DD HH24:MI:SS.US\'::varchar),\n\t\t\tTO_TIMESTAMP(\'2021-06-01 15:59:59.210732\', \'YYYY-MM-DD HH24:MI:SS.US\'::varchar),\n\t\t\tnull,\n\t\t\tnull\n\t\tFROM (\n\t\t\tSELECT MAX("fmc_end_lw_timestamp") as "fmc_begin_lw_timestamp" \n\t\t\tFROM "moto_fmc"."fmc_loading_history" \n\t\t\tWHERE "src_bk" = \'airflow\' \n\t\t\tAND "success_flag" = 1\n\t\t\tAND "load_cycle_id" < 35\n\t\t) SRC_WINDOW\n\t\tWHERE NOT EXISTS(SELECT 1 FROM "moto_fmc"."fmc_loading_history" WHERE "load_cycle_id" = 35)', 'TRUNCATE TABLE "airflow_mtd"."load_cycle_info" ', 'INSERT INTO "airflow_mtd"."load_cycle_info"("load_cycle_id","load_date") \n\t\t\tSELECT 35,TO_TIMESTAMP(\'2021-06-01 16:00:00.000000\', \'YYYY-MM-DD HH24:MI:SS.US\'::varchar)', 'TRUNCATE TABLE "airflow_mtd"."fmc_loading_window_table" ', 'INSERT INTO "airflow_mtd"."fmc_loading_window_table"("fmc_begin_lw_timestamp","fmc_end_lw_timestamp") \n\t\t\tSELECT "fmc_begin_lw_timestamp" + -15 * interval\'1 minute\', TO_TIMESTAMP(\'2021-06-01 16:00:00.000000\', \'YYYY-MM-DD HH24:MI:SS.US\'::varchar)\n\t\t\tFROM (\n\t\t\t\tSELECT MAX("fmc_end_lw_timestamp") as "fmc_begin_lw_timestamp" \n\t\t\t\tFROM "moto_fmc"."fmc_loading_history" \n\t\t\t\tWHERE "src_bk" = \'airflow\' \n\t\t\t\tAND "success_flag" = 1\n\t\t\t\tAND "load_cycle_id" < 35\n\t\t\t) SRC_WINDOW']
[2021-06-01 18:00:18,097] {base.py:78} INFO - Using connection to: id: test_dv. Host: jdbc:postgresql://localhost:5432/test_dv_stijn, Port: None, Schema: , Login: postgres, Password: ***, extra: {'extra__jdbc__drv_path': '/home/stijn/airflow/jdbc/postgresql-9.4.1212.jar', 'extra__jdbc__drv_clsname': 'org.postgresql.Driver', 'extra__google_cloud_platform__project': '', 'extra__google_cloud_platform__key_path': '', 'extra__google_cloud_platform__keyfile_dict': '', 'extra__google_cloud_platform__scope': '', 'extra__google_cloud_platform__num_retries': 5, 'extra__grpc__auth_type': '', 'extra__grpc__credential_pem_file': '', 'extra__grpc__scopes': '', 'extra__yandexcloud__service_account_json': '', 'extra__yandexcloud__service_account_json_path': '', 'extra__yandexcloud__oauth': '', 'extra__yandexcloud__public_ssh_key': '', 'extra__yandexcloud__folder_id': '', 'extra__kubernetes__in_cluster': False, 'extra__kubernetes__kube_config': '', 'extra__kubernetes__namespace': ''}
[2021-06-01 18:00:18,530] {local_task_job.py:151} INFO - Task exited with return code 1
`

### LINKED ISSUE #18482: No SQL error shown when using the JDBC operator
### Apache Airflow version

2.2.0b1 (beta snapshot)

### Operating System

Debian buster

### Versions of Apache Airflow Providers

_No response_

### Deployment

Astronomer

### Deployment details

astro dev start
Dockerfile - https://gist.github.com/jyotsa09/267940333ffae4d9f3a51ac19762c094#file-dockerfile

### What happened

Both connection pointed to the same database where table "Person" doesn't exist.

When there is a SQL failure in PostgressOperator then logs are saying - 
```
psycopg2.errors.UndefinedTable: relation "person" does not exist
LINE 1: Select * from Person
```

When there is a SQL failure in JDBCOperator then logs are just saying -
```
[2021-09-23, 19:06:22 UTC] {local_task_job.py:154} INFO - Task exited with return code
```

### What you expected to happen

JDBCOperator should show "Task failed with exception" or something similar to Postgress.

### How to reproduce

Use this dag - https://gist.github.com/jyotsa09/267940333ffae4d9f3a51ac19762c094 
(Connection extras is in the gist.)

### Anything else

Related: https://github.com/apache/airflow/issues/16564

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by hubert-pietron**
closes: #16295 
closes: #18482

JDBC operator could not log errors when task failed. For specific type of error (jpype.java.sql.SQLException, jaydebeapi.DatabaseError, jaydebeapi.InterfaceError) this line https://github.com/apache/airflow/blob/0a2d0d1ecbb7a72677f96bc17117799ab40853e0/airflow/models/taskinstance.py#L1733
would produce AttributeError
```
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING - Traceback (most recent call last):
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/opt/***/***/models/taskinstance.py", line 1726, in handle_failure
    self.log.error("Task failed with exception", exc_info=error.__traceback__)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/usr/local/lib/python3.8/logging/__init__.py", line 1475, in error
    self._log(ERROR, msg, args, **kwargs)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/usr/local/lib/python3.8/logging/__init__.py", line 1589, in _log
    self.handle(record)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/usr/local/lib/python3.8/logging/__init__.py", line 1599, in handle
    self.callHandlers(record)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/usr/local/lib/python3.8/logging/__init__.py", line 1661, in callHandlers
    hdlr.handle(record)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/usr/local/lib/python3.8/logging/__init__.py", line 950, in handle
    rv = self.filter(record)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/usr/local/lib/python3.8/logging/__init__.py", line 811, in filter
    result = f.filter(record)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/opt/***/***/utils/log/secrets_masker.py", line 169, in filter
    self._redact_exception_with_context(exc)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/opt/***/***/utils/log/secrets_masker.py", line 150, in _redact_exception_with_context
    exception.args = (self.redact(v) for v in exception.args)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING - AttributeError: can't set attribute
``` 
Catching that error allow to see meaningful log when JDBC task fails.

example DAG
```
from datetime import datetime, timedelta
from airflow import DAG
from airflow.providers.jdbc.operators.jdbc import JdbcOperator

with DAG(
    dag_id='jdbc_with_error',
    schedule_interval='0 0 * * *',
    start_date=datetime(2021, 1, 1),
    dagrun_timeout=timedelta(minutes=60),
    tags=['example'],
    catchup=False,
) as dag:

    start = JdbcOperator(
        task_id='test_task',
        sql='select 1 x y z',
        jdbc_conn_id='my_jdbc_connection',
        autocommit=True,
    )
```
log after change
```
*** Reading local file: /root/airflow/logs/jdbc_with_error/test_task/2022-02-12T15:38:46.534291+00:00/1.log
[2022-02-12, 15:38:47 UTC] {taskinstance.py:1052} INFO - Dependencies all met for <TaskInstance: jdbc_with_error.test_task manual__2022-02-12T15:38:46.534291+00:00 [queued]>
[2022-02-12, 15:38:47 UTC] {taskinstance.py:1052} INFO - Dependencies all met for <TaskInstance: jdbc_with_error.test_task manual__2022-02-12T15:38:46.534291+00:00 [queued]>
[2022-02-12, 15:38:47 UTC] {taskinstance.py:1249} INFO - 
--------------------------------------------------------------------------------
[2022-02-12, 15:38:47 UTC] {taskinstance.py:1250} INFO - Starting attempt 1 of 1
[2022-02-12, 15:38:47 UTC] {taskinstance.py:1251} INFO - 
--------------------------------------------------------------------------------
[2022-02-12, 15:38:47 UTC] {taskinstance.py:1270} INFO - Executing <Task(JdbcOperator): test_task> on 2022-02-12 15:38:46.534291+00:00
[2022-02-12, 15:38:47 UTC] {standard_task_runner.py:52} INFO - Started process 7755 to run task
[2022-02-12, 15:38:47 UTC] {standard_task_runner.py:76} INFO - Running: ['***', 'tasks', 'run', 'jdbc_with_error', 'test_task', 'manual__2022-02-12T15:38:46.534291+00:00', '--job-id', '245', '--raw', '--subdir', 'DAGS_FOLDER/example_dag.py', '--cfg-path', '/tmp/tmp7ryzrpgx', '--error-file', '/tmp/tmp0u3p6q8m']
[2022-02-12, 15:38:47 UTC] {standard_task_runner.py:77} INFO - Job 245: Subtask test_task
[2022-02-12, 15:38:47 UTC] {logging_mixin.py:115} INFO - Running <TaskInstance: jdbc_with_error.test_task manual__2022-02-12T15:38:46.534291+00:00 [running]> on host 464ba29d61dc
[2022-02-12, 15:38:47 UTC] {logging_mixin.py:115} WARNING - /opt/***/***/models/taskinstance.py:830 DeprecationWarning: Passing 'execution_date' to 'XCom.clear()' is deprecated. Use 'run_id' instead.
[2022-02-12, 15:38:47 UTC] {logging_mixin.py:115} WARNING - /usr/local/lib/python3.8/site-packages/sqlalchemy/sql/coercions.py:518 SAWarning: Coercing Subquery object into a select() for use in IN(); please pass a select() construct explicitly
[2022-02-12, 15:38:48 UTC] {taskinstance.py:1442} INFO - Exporting the following env vars:
AIRFLOW_CTX_DAG_OWNER=***
AIRFLOW_CTX_DAG_ID=jdbc_with_error
AIRFLOW_CTX_TASK_ID=test_task
AIRFLOW_CTX_EXECUTION_DATE=2022-02-12T15:38:46.534291+00:00
AIRFLOW_CTX_DAG_RUN_ID=manual__2022-02-12T15:38:46.534291+00:00
[2022-02-12, 15:38:48 UTC] {jdbc.py:68} INFO - Executing: select 1 x y z
[2022-02-12, 15:38:48 UTC] {base.py:70} INFO - Using connection to: id: my_jdbc_connection. Host: jdbc:sqlite:elo, Port: None, Schema: , Login: , Password: None, extra: {'extra__jdbc__drv_clsname': 'org.sqlite.JDBC', 'extra__jdbc__drv_path': '/opt/***/sqlite-jdbc-3.36.0.3.jar'}
[2022-02-12, 15:38:48 UTC] {dbapi.py:213} INFO - Running statement: select 1 x y z, parameters: None
[2022-02-12, 15:38:48 UTC] {taskinstance.py:1738} ERROR - org.sqlite.SQLiteException: [SQLITE_ERROR] SQL error or missing database (near "y": syntax error)
[2022-02-12, 15:38:48 UTC] {taskinstance.py:1278} INFO - Marking task as FAILED. dag_id=jdbc_with_error, task_id=test_task, execution_date=20220212T153846, start_date=20220212T153847, end_date=20220212T153848
[2022-02-12, 15:38:48 UTC] {local_task_job.py:162} INFO - Task exited with return code 1
[2022-02-12, 15:38:48 UTC] {local_task_job.py:279} INFO - 0 downstream tasks scheduled from follow-on schedule check
[2022-02-12, 15:38:48 UTC] {logging_mixin.py:115} WARNING - /opt/***/***/models/dag.py:1068 SADeprecationWarning: Query.value() is deprecated and will be removed in a future release.  Please use Query.with_entities() in combination with Query.scalar() (deprecated since: 1.4)
```

**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Review by o-nikolas — COMMENTED** [review]
It looks like the exception that's happening during the handling of the first exception is actually Airflow filtering code (the secret masker specifically):
```[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING -   File "/opt/***/***/utils/log/secrets_masker.py", line 150, in _redact_exception_with_context
    exception.args = (self.redact(v) for v in exception.args)
[2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING - AttributeError: can't set attribute
```

I think this code should be patched to handle whatever edge case is occurring here, rather than the caller just catching the exception and logging it another way to work around it.

It looks like @potiuk was the last to touch the `_redact_exception_with_context` method, perhaps he has more context.

**Comment by potiuk** [comment]
> It looks like the exception that's happening during the handling of the first exception is actually Airflow filtering code (the secret masker specifically):
> 
> ```
>     exception.args = (self.redact(v) for v in exception.args)
> [2022-02-12, 09:27:16 UTC] {logging_mixin.py:115} WARNING - AttributeError: can't set attribute
> ```
> 
> I think this code should be patched to handle whatever edge case is occurring here, rather than the caller just catching the exception and logging it another way to work around it.
> 
> It looks like @potiuk was the last to touch the `_redact_exception_with_context` method, perhaps he has more context.

Yeah. I refectored a code that had literally this comment:

>  # I'm not sure if this is a good idea!

Apparently it was not.

I think the right approach is to make the same try/except but in the place where redact is mdone @hubert-pietron

Just find:
```
        exception.args = (self.redact(v) for v in exception.args)
```

and surround it with the:

```
try:
        exception.args = (self.redact(v) for v in exception.args)
except AttributeError:
       pass
```

The problem is that in some exceptions the args are "read property" and you cannot replace them.

**Comment by hubert-pietron** [comment]
Thanks for response! I pushed changes like you suggested @potiuk

**Review by o-nikolas — APPROVED** [review]

File: airflow/utils/log/secrets_masker.py [hunk 1]
```diff
@@ -147,7 +147,10 @@ def _record_attrs_to_ignore(self) -> Iterable[str]:
         return frozenset(record.__dict__).difference({'msg', 'args'})
 
     def _redact_exception_with_context(self, exception):
-        exception.args = (self.redact(v) for v in exception.args)
+        try:
+            exception.args = (self.redact(v) for v in exception.args)
+        except AttributeError:
```
Comment (uranusjr) [inline airflow/utils/log/secrets_masker.py:154]: Add a comment explaining why this is done?

File: airflow/utils/log/secrets_masker.py [hunk 2]
```diff
@@ -147,7 +147,12 @@ def _record_attrs_to_ignore(self) -> Iterable[str]:
         return frozenset(record.__dict__).difference({'msg', 'args'})
 
     def _redact_exception_with_context(self, exception):
-        exception.args = (self.redact(v) for v in exception.args)
+        # In case, when exception has "read only" properties, we need
+        # to catch an AttributeError to log properly
```
Comment (uranusjr) [inline airflow/utils/log/secrets_masker.py]:
```suggestion
        # Exception class may not be modifiable (e.g. declared by an
        # extension module such as JDBC).
```

**Review by uranusjr — APPROVED** [review]

**Review by potiuk — APPROVED** [review]

**Pull request merged**